wavefront/wavefront.go

```diff
@@ -48,25 +52,74 @@ type (
 
 	// WavefrontAdapter supports metric template.
 	WavefrontAdapter struct {
-		listener            net.Listener
-		server              *grpc.Server
-		reporterInitialized bool
+		listener net.Listener
+		server   *grpc.Server
+		reporter wf.WavefrontMetricsReporter
 	}
 )
 
 // ensure that WavefrontAdapter implements the HandleMetricServiceServer interface.
 var _ metric.HandleMetricServiceServer = &WavefrontAdapter{}
 
 // createWavefrontReporter creates a reporter that periodically flushes metrics to Wavefront.
-func createWavefrontReporter(cfg *config.Params) {
-	hostTags := map[string]string{"source": cfg.Source}
+func (wa *WavefrontAdapter) createWavefrontReporter(cfg *config.Params) {
+
 	if direct := cfg.GetDirect(); direct != nil {
-		go wf.WavefrontDirect(metrics.DefaultRegistry, cfg.FlushInterval, hostTags, cfg.Prefix, direct.Server, direct.Token)
+		directCfg := &senders.DirectConfiguration{
+			Server:               direct.Server,
+			Token:                direct.Token,
+			FlushIntervalSeconds: int(cfg.FlushInterval.Seconds()),
+			BatchSize:            10000,
+			MaxBufferSize:        50000,
+		}
+		sender, err := senders.NewDirectSender(directCfg)
+		if err != nil {
+			panic(err)
+		}
+		wa.reporter = wf.NewReporter(
+			sender,
+			application.New("wavefront-istio-adapter", "wavefront-istio-adapter"),
+			wf.Source(cfg.Source),
+			wf.Prefix(cfg.Prefix),
+			wf.LogErrors(true),
+		)
 	} else if proxy := cfg.GetProxy(); proxy != nil {
 		addr, _ := net.ResolveTCPAddr("tcp", proxy.Address)
-		go wf.WavefrontProxy(metrics.DefaultRegistry, cfg.FlushInterval, hostTags, cfg.Prefix, addr)
-	}
 
+		// extract proxy ip and port from address
+		proxyInfo := strings.Split(addr.String(), ":")
+
+		// address must be in the form <proxyhost:port>
+		if len(proxyInfo) != 2 {
+			panic("Proxy Address and/or port number is missing.")
+		}
+
+		// numeric port number expected
+		portNum, err := strconv.Atoi(proxyInfo[1])
+		if err != nil {
+			panic(err)
+		}
+
+		proxyCfg := &senders.ProxyConfiguration{
+			Host:                 proxyInfo[0],
+			MetricsPort:          portNum,
+			FlushIntervalSeconds: int(cfg.FlushInterval.Seconds()),
+		}
+
+		sender, err := senders.NewProxySender(proxyCfg)
+		if err != nil {
+			panic(err)
+		}
+
+		wa.reporter = wf.NewReporter(
+			sender,
+			application.New("wavefront-istio-adapter", "wavefront-istio-adapter"),
+			wf.Source(cfg.Source),
+			wf.Prefix(cfg.Prefix),
+			wf.LogErrors(true),
+		)
```

Comment:
Looks like this part is very similar in the direct block. Can you simplify this by splitting this reporter/sender initialization logic into separate methods?

You can probably build senders for direct and proxy in 2 separate methods. Then you can pass those into the reporter here.

Reply:
Added two separate methods and simplified the logic.